### PR TITLE
Base for the Input system.

### LINKED ISSUE #9: Input

### PR DESCRIPTION
**Pull request by Fusioon**

### CODE REVIEW

**Comment by RogueMacro** [comment]
@Fusioon We wouldn't need a keymap if we just assigned the keycodes to their glfw equivalent. Since we are only using glfw for now and not the win32 api that should be fine. What do you think?

File: Engine/src/Application.bf [hunk 1]
```diff
@@ -51,6 +53,15 @@ namespace SteelEngine
 		{
 			var dispatcher = scope EventDispatcher(event);
 			dispatcher.Dispatch<WindowCloseEvent>(scope => OnWindowClose);
+			dispatcher.Dispatch<KeyPressedEvent>(scope => OnKeyPressed);
+			dispatcher.Dispatch<KeyReleasedEvent>(scope => OnKeyRelease);
+			dispatcher.Dispatch<MouseButtonPressedEvent>(scope => OnMouseButtonPressed);
+			dispatcher.Dispatch<MouseButtonReleasedEvent>(scope => OnMouseButtonReleased);
+
+			if(event.EventType == .WindowLostFocus)
```
Comment (jamesaorson) [inline Engine/src/Application.bf]: Add a space between if and (

File: Engine/src/Input/GLFWKeyMapper.bf [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+			(.Num1, .Alpha2),
+			(.Num1, .Alpha3),
+			(.Num1, .Alpha4),
+			(.Num1, .Alpha5),
+			(.Num1, .Alpha6),
+			(.Num1, .Alpha7),
+			(.Num1, .Alpha8),
+			(.Num1, .Alpha9),
+
+			(.Smicolon, .Semicolon),
```
Comment (jamesaorson) [inline Engine/src/Input/GLFWKeyMapper.bf:29]: Is it really Smicolon in Glfw?

File: Engine/src/Input/GLFWKeyMapper.bf [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+			(.RightControl, .RightControl),
+			(.RightAlt, .RightAlt),
+			(.RightSuper, .LeftWindows),	// @TODO
+			(.Menu, .Menu),
+		} ~ delete _;
+
+		public static KeyCode MapKeyboardKey(GlfwInput.Key key)
+		{
+			KeyCode kc;
+			if(_keyCodeMap.TryGetValue(key, out kc))
```
Comment (jamesaorson) [inline Engine/src/Input/GLFWKeyMapper.bf]: Add a space between if and (

File: Engine/src/Input/Input.bf [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		static var _lastUpdateState = KeyStatus[KeyCode.MAX]();
+
+		static void KeyEvent(KeyCode kc, KeyEvent ke)
+		{
+			_accumulatedEvents[kc.Underlying] |= ke;
+		}
+
+		static void Update()
+		{
+			for(int i = 0, let count = _accumulatedEvents.Count; i < count; i++)
```
Comment (jamesaorson) [inline Engine/src/Input/Input.bf]: Add a space between for and (

File: Engine/src/Input/Input.bf [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+			{
+				let event = _accumulatedEvents[i];
+
+				_lastUpdateState[i] &= ~(.Down | .Up);	// Clear up / down
+				if(event.HasFlag(.Down))
```
Comment (jamesaorson) [inline Engine/src/Input/Input.bf]: Add a space between if and (

File: Engine/src/Input/Input.bf [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+				{
+					_lastUpdateState[i] = .Down | .Hold;
+				}
+
+				if(event.HasFlag(.Up))
```
Comment (jamesaorson) [inline Engine/src/Input/Input.bf]: Add a space between if and (

**Review by jamesaorson — CHANGES_REQUESTED** [review]

File: Engine/src/Input/GLFWKeyMapper.bf (hunk 2 above)
Comment (Fusioon) [reply Engine/src/Input/GLFWKeyMapper.bf:29]: Yes it is.

**Review by jamesaorson — APPROVED** [review]

**Comment by jamesaorson** [comment]
> @Fusioon We wouldn't need a keymap if we just assigned the keycodes to their glfw equivalent. Since we are only using glfw for now and not the win32 api that should be fine. What do you think?

We want to abstract outselves away from whatever input handler we are using. That way, developers only see Steel's API. They aren't dealing with the underlying implementation. Say we switched to use SDL over GLFW, now Steel developers would have to change their uses all over the code base. By just mapping to a standard that Steel will uphold, everyone is safe even if we switch implementations.

**Pull request merged**